File: app/src/api/catalogue.ts
import { invoke } from "@tauri-apps/api/core";

export interface Category {
  id: number;
  name: string;
  description: string;
  color: string;
  created_at: string;
}

export interface SaveCategoryPayload {
  name: string;
  description: string;
  color: string;
}

export interface CatalogueItem {
  id: number;
  item_type: "product" | "service";
  category_id: number | null;
  category_name: string | null;
  name: string;
  description: string;
  unit_price: number;
  unit: string;
  tva_applicable: boolean;
  active: boolean;
  created_at: string;
  updated_at: string;
}

export interface SaveCataloguePayload {
  item_type: "product" | "service";
  category_id: number | null;
  name: string;
  description: string;
  unit_price: number;
  unit: string;
  tva_applicable: boolean;
}

export async function getCategories(): Promise<Category[]> {
  return invoke<Category[]>("get_categories");
}

export async function createCategory(payload: SaveCategoryPayload): Promise<Category> {
  return invoke<Category>("create_category", { payload });
}

export async function updateCategory(id: number, payload: SaveCategoryPayload): Promise<Category> {
  return invoke<Category>("update_category", { id, payload });
}

export async function deleteCategory(id: number): Promise<void> {
  return invoke<void>("delete_category", { id });
}

export async function getCatalogue(activeOnly = true): Promise<CatalogueItem[]> {
  return invoke<CatalogueItem[]>("get_catalogue", { activeOnly });
}

export async function createCatalogueItem(
  payload: SaveCataloguePayload,
): Promise<CatalogueItem> {
  return invoke<CatalogueItem>("create_catalogue_item", { payload });
}

export async function updateCatalogueItem(
  id: number,
  payload: SaveCataloguePayload,
): Promise<CatalogueItem> {
  return invoke<CatalogueItem>("update_catalogue_item", { id, payload });
}

export async function toggleCatalogueItem(id: number): Promise<CatalogueItem> {
  return invoke<CatalogueItem>("toggle_catalogue_item", { id });
}

File: app/src-tauri/src/models/catalogue.rs
use serde::{Deserialize, Serialize};


#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Category {
    pub id: i64,
    pub name: String,
    pub description: String,
    pub color: String,
    pub created_at: String,
}

#[derive(Debug, Deserialize)]
pub struct SaveCategoryPayload {
    pub name: String,
    pub description: String,
    pub color: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CatalogueItem {
    pub id: i64,
    pub item_type: String, // "product" | "service"
    pub category_id: Option<i64>,
    pub category_name: Option<String>,
    pub name: String,
    pub description: String,
    pub unit_price: i64,
    pub unit: String,
    pub tva_applicable: bool,
    pub active: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Deserialize)]
pub struct SaveCataloguePayload {
    pub item_type: String,
    pub category_id: Option<i64>,
    pub name: String,
    pub description: String,
    pub unit_price: i64,
    pub unit: String,
    pub tva_applicable: bool,
}
